### ISSUE TITLE
REST server crashes with NullPointerException

### ISSUE THREAD

**Issue by marcomicera** [fasten-project/fasten#273]
## Describe the bug

Upon building and launching the REST server, it crashes with a `NullPointerException`.

## To Reproduce

1. Install the required projects:
    ```shell
    mvn clean install -Dmaven.test.skip=true --projects core,:restapi-plugin --also-make
    ```
1. Start the REST server:
    ```shell
    PGPASSWORD=... mvn spring-boot:run -f analyzer/restapi-plugin/pom.xml
    ```

## Expected behavior

The REST server should start accepting HTTP requests.

## Logs

<details>
<summary>Click to expand</summary>

```
INFO    8156 --- [           main] e.f.a.restapiplugin.mvn.RestApplication  : Started RestApplication in 6.873 seconds (JVM running for 7.298)
DEBUG   8156 --- [           main] o.s.web.client.RestTemplate              : HTTP GET http://localhost:8080/__INTERNAL__/packages/0/directedgraph
DEBUG   8156 --- [           main] o.s.web.client.RestTemplate              : Accept=[text/plain, application/json, application/*+json, */*]
INFO    8156 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
INFO    8156 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
DEBUG    8156 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Detected StandardServletMultipartResolver
DEBUG   8156 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : enableLoggingRequestDetails='false': request parameters and headers will be masked to prevent unsafe logging of potentially sensitive data
INFO    8156 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 8 ms
DEBUG   8156 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : GET "/__INTERNAL__/packages/0/directedgraph", parameters={}
WARN    8156 --- [nio-8080-exec-1] e.f.c.m.utils.DependencyGraphUtilities   : Graph at null is incomplete
ERROR   8156 --- [nio-8080-exec-1] e.f.a.r.m.a.i.ResolutionApiServiceImpl   : Error constructing dependency graph resolver

java.lang.NullPointerException: null
	at eu.fasten.core.maven.DependencyGraphBuilder.getDependencyList(DependencyGraphBuilder.java:65) ~[core-0.0.1-SNAPSHOT.jar:na]
	at eu.fasten.core.maven.DependencyGraphBuilder.buildDependencyGraph(DependencyGraphBuilder.java:160) ~[core-0.0.1-SNAPSHOT.jar:na]
	at eu.fasten.core.maven.utils.DependencyGraphUtilities.buildDependencyGraphFromScratch(DependencyGraphUtilities.java:191) ~[core-0.0.1-SNAPSHOT.jar:na]
	at eu.fasten.core.maven.GraphMavenResolver.buildDependencyGraph(GraphMavenResolver.java:525) ~[core-0.0.1-SNAPSHOT.jar:na]
	at eu.fasten.analyzer.restapiplugin.mvn.api.impl.ResolutionApiServiceImpl.<init>(ResolutionApiServiceImpl.java:54) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[na:na]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:204) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:87) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1310) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1213) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:556) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:516) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:324) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1307) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1227) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:884) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:788) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:227) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1356) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1203) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:556) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:516) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:324) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:322) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.method.HandlerMethod.createWithResolvedBean(HandlerMethod.java:336) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.handler.AbstractHandlerMethodMapping.getHandlerInternal(AbstractHandlerMethodMapping.java:368) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.mvc.method.RequestMappingInfoHandlerMapping.getHandlerInternal(RequestMappingInfoHandlerMapping.java:110) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.mvc.method.RequestMappingInfoHandlerMapping.getHandlerInternal(RequestMappingInfoHandlerMapping.java:59) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.handler.AbstractHandlerMapping.getHandler(AbstractHandlerMapping.java:396) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.getHandler(DispatcherServlet.java:1234) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1016) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:626) ~[tomcat-embed-core-9.0.39.jar:4.0.FR]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:733) ~[tomcat-embed-core-9.0.39.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.10.RELEASE.jar:5.2.10.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1590) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-9.0.39.jar:9.0.39]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
```
</details>

## Environment (please complete the following information):

- OS:
    ```shell
    $ sw_vers
    ProductName:	macOS
    ProductVersion:	11.0.1
    BuildVersion:	20B29
    ```
- Java version:
    ```shell
    $ java --version
    openjdk 11.0.2 2019-01-15
    OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
    OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
    ```

## Additional context

The problem might be caused by the following edits I've made: since I don't know how to set [these newly added Spring Boot properties](https://github.com/fasten-project/fasten/blob/15d0854a44cfd65620dcb982e354b5977296b4fa/analyzer/restapi-plugin/src/main/resources/application.properties#L8-L12), I thought that by commenting [these unused fields](https://github.com/fasten-project/fasten/blob/15d0854a44cfd65620dcb982e354b5977296b4fa/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/KnowledgeBaseConnector.java#L75-L91) and [these unused methods](https://github.com/fasten-project/fasten/blob/15d0854a44cfd65620dcb982e354b5977296b4fa/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/KnowledgeBaseConnector.java#L93-L152), everything would have been working just fine.

Does my IDE (IntelliJ) erroneously detect these fields and methods as unused?
Perhaps they are being used by Spring boot, hence necessary for everything to work?

**Comment by MihhailSokolov** [comment]
Judging by the logs, it looks like the problem is that the location for the dependency graph is not properly specified. In the `application.properties` you should either specify the actual values for the attributes when you run it locally or you should pass those values as the arguments.
For the first option, here is the `application.properties` file I use when running REST API locally:
```
logging.level.org.springframework.web=DEBUG
spring.mvc.async.request-timeout=60000
server.tomcat.connection-timeout=60000
server.port=8080
server.tomcat.threads.max=20
kb.user=postgres
kb.url=jdbc:postgresql://localhost:5432/postgres
kb.depgraph.path=../depgraph/mvn_dep_graph
kb.graphdb.path=../graphDB
lima.rcg.url=https://lima.ewi.tudelft.nl/
kafka.address=localhost:9092
kafka.output.topic=fasten.mvn.priority.ingest
```

For the second option, you should add an argument for each property in the `application.properties`: `lima.rcg.url=${rcg.url}` -> `--rcg.url='https://lima.ewi.tudelft.nl/'`. The full list of arguments can be found in the [REST API deployment file](https://github.com/fasten-project/k8s-deployments/blob/master/fasten/rest-api/restapi-deployment.yaml#L19)

**Comment by amphioxis** [comment]
Running exactly that file, recompile everything with ```mvn clean install -Dmaven.test.skip=true --projects core --also-make && PGPASSWORD=... mvn clean install -Dmaven.test.skip=true spring-boot:run -f analyzer/restapi-plugin/pom.xml```

results in another error:

```Couldn't connect to the Graph Database

java.lang.RuntimeException: org.rocksdb.RocksDBException: While opening a file for sequentially reading: ../graphDB/CURRENT: No such file or directory```

**Comment by MihhailSokolov** [comment]
> Running exactly that file, recompile everything with
> `mvn clean install -Dmaven.test.skip=true --projects core --also-make && PGPASSWORD=... mvn clean install -Dmaven.test.skip=true spring-boot:run -f analyzer/restapi-plugin/pom.xml`
> 
> results in another error:
> 
> `Couldn't connect to the Graph Database`
> `java.lang.RuntimeException: org.rocksdb.RocksDBException: While opening a file for sequentially reading: ../graphDB/CURRENT: No such file or directory`

Well, that's because you don't have a graph database set up locally (at least at that certain location).
For the graph database, we use RocksDB, so it does not require any additional installation. I think creating that `graphDB` folder should do it.
You will most likely run into a similar problem with the dependency graph as you probably don't have it locally too. The `../depgraph/mvn_dep_graph` actually refers to 2 files - `../depgraph/mvn_dep_graph.nodes` and `../depgraph/mvn_dep_graph.edges`. So you need to have these two files at the specified location. You can copy these from Monster `/mnt/fasten/mvn-depgraph/mvn_dep_graph.*`.
Let me know if you run into more troubles because setting up the REST API to run locally can be a bit challenging :)

**Comment by amphioxis** [comment]
Right now I am running into an ```Connection to node -1 (localhost/127.0.0.1:9092) could not be established. Broker may not be available.``` error. caused by Kafka NetworkClient.
I also changed the user to fasten and used his password, as I don't have access to the password for the postgres user. My file looks like this right now:

```logging.level.org.springframework.web=DEBUG
spring.mvc.async.request-timeout=60000
server.tomcat.connection-timeout=60000
server.port=8080
server.tomcat.threads.max=20
kb.user=fasten
kb.url=jdbc:postgresql:fasten_python
kb.depgraph.path=../depgraph/mvn_dep_graph
kb.graphdb.path=../graphDB
lima.rcg.url=https://lima.ewi.tudelft.nl/
kafka.address=localhost:9092
kafka.output.topic=fasten.mvn.priority.ingest```

**Comment by MihhailSokolov** [comment]
> Right now I am running into an ```Connection to node -1 (localhost/127.0.0.1:9092) could not be established. Broker may not be available.``` error caused by Kafka NetworkClient.

Do you have Apache Kafka installed and running? Looks like the REST API cannot connect to Kafka

**Comment by marcomicera** [comment]
Just wondering: why does the REST server need to connect to Kafka? Are there any new endpoints that trigger other components?

**Comment by MihhailSokolov** [comment]
> Just wondering: why does the REST server need to connect to Kafka? Are there any new endpoints that trigger other components?

Yes, about 2 months ago I added a [lazy ingestion system](https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects#lazy-ingestion) - when a client requests a package version that is not in the database, this package version's Maven coordinates (if they are valid and haven't been ingested yet) are written to the input Kafka topic for our processing pipeline, thus after a bit of time the package version will be injected in the database and the client can request it again.

**Comment by marcomicera** [comment]
Ok, I understand. I guess @amphioxis doesn't need these endpoints as he just needs to extend the REST API for Python packages. What about creating an option/flag/whatever that prevents these endpoints to be created, so that he won't need to spawn a Kafka instance locally, and so on? This could be a hidden dev option somewhere... how does it sound?

**Comment by MihhailSokolov** [comment]
> Ok, I understand. I guess @amphioxis doesn't need these endpoints as he just needs to extend the REST API for Python packages. What about creating an option/flag/whatever that prevents these endpoints to be created, so that he won't need to spawn a Kafka instance locally, and so on? This could be a hidden dev option somewhere... how does it sound?

I guess then we would have to run two separate instances of the REST API (one with Kafka connection and one without). Connection to Kafka is established during startup (same way as connection to the database), so I am not sure it is possible to enable it only for some of the endpoints.

**Comment by marcomicera** [comment]
> I guess then we would have to run two separate instances of the REST API (one with Kafka connection and one without).

No my idea is to disable that part only during development (@amphioxis could be using a completely separate Kafka instance), and then we wouldn't disable that part in prod. I was just wondering whether he could push this "disable lazy ingestion endpoints" flag, setting its default value to `false`.

**Comment by MihhailSokolov** [comment]
> > I guess then we would have to run two separate instances of the REST API (one with Kafka connection and one without).
> 
> No my idea is to disable that part only during development (@amphioxis could be using a completely separate Kafka instance), and then we wouldn't disable that part in prod. I was just wondering whether he could push this "disable lazy ingestion endpoints" flag, setting its default value to `false`.

Yes, it is possible to add this kind of flag. However, I think it would make more sense to keep it as it is. Python endpoints can make use of lazy ingestion as well because all the infrastructure for it is ready. The partners are quite actively using lazy ingestion for Java, so it would be handy to have it for Python too. Plus, just installing Kafka seems more logical than adding some flag that will have to be removed later for production.

**Comment by marcomicera** [comment]
> The partners are quite actively using lazy ingestion for Java, so it would be handy to have it for Python too. Plus, just installing Kafka seems more logical than adding some flag that will have to be removed later for production.

Sorry for the late reply! Yes, I understand.
I guess the only service that @amphioxis is missing in order to set everything up is the graph database (he already has access to the PostgreSQL database): can we have access to it?
In our docker-compose deployment repository is only see a `fasten-graph-db-extension-java` service, which, I guess, it's just an extension...

Also, are there any other services we need to start the REST API, besides Kafka, PostgreSQL, and the graph DB? Am I missing something?

**Comment by MihhailSokolov** [comment]
> I guess the only service that @amphioxis is missing in order to set everything up is the graph database (he already has access to the PostgreSQL database): can we have access to it?
> In our docker-compose deployment repository is only see a `fasten-graph-db-extension-java` service, which, I guess, it's just an extension...

Yes, you should have access to the graph database. Graph db is just a binary key-value store, so everything is stored in files on Monster. It is located in `/mnt/fasten/graphdb`. I am not sure what your docker-compose `fasten-graph-db-extension-java` does, but I would assume it creates a local instance of the graph db in the specified directory and writes there data received (using Kafka) from metadata plugin.

> Also, are there any other services we need to start the REST API, besides Kafka, PostgreSQL, and the graph DB? Am I missing something?

I think you might need serialized dependency graph as well, but it is not required. I can also be reconstructed from the postgres database. It will be constructed and serialized to the path specified in the application.properties` file. Otherwise, you could copy the existing serialized dep graph from Monster. It is located in `/mnt/fasten/mvn-depgraph`.

Let me know if anything is still unclear.

**Comment by amphioxis** [comment]
Closed since using the [docker-fasten-deployment](https://github.com/fasten-project/fasten-docker-deployment) solved the issue.

**Comment by marcomicera** [comment]
Has this been fixed? Can we close this? @amphioxis

**Comment by amphioxis** [comment]
I switched to the [fasten-docker-deployment](https://github.com/fasten-project/fasten-docker-deployment) which runs all the requirements like `kafka` and the `graph database` at once without problems. However, the last time I ran the `FASTEN REST` server locally, it crashed with a `java heap space` error while deserializing the Call Graph file I downloaded from the server at `/mnt/fasten/graphdb`. But I'm not sure if that was caused because of the missing `kafka` (I switched the connection off) or if that was/is a "real" bug. So I close this now, because my problems are solved.
